src/android/UpdateHashUtils.java

```diff
@@ -0,0 +1,73 @@
+package com.microsoft.cordova;
+
+import android.app.Activity;
+import android.content.res.AssetManager;
+
+import org.json.JSONArray;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.security.DigestInputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Collections;
+
+/**
+ * Utilities class used for native operations related to calculating hashes of update contents.
+ */
+public class UpdateHashUtils {
+
+    public static String getBinaryHash(Activity activity) throws IOException {
+        ArrayList<String> manifestEntries = new ArrayList<String>();
+        addFolderEntriesToManifest(manifestEntries, "www", activity.getAssets());
+        Collections.sort(manifestEntries);
+        JSONArray manifestJSONArray = new JSONArray();
+        for (String manifestEntry : manifestEntries) {
+            manifestJSONArray.put(manifestEntry);
+        }
+
+        // The JSON serialization turns path separators into "\/", e.g. "www\/images\/image.png"
+        String manifestString = manifestJSONArray.toString().replace("\\/", "/");
+        return computeHash(new ByteArrayInputStream(manifestString.getBytes()));
+    }
+
+    private static void addFolderEntriesToManifest(ArrayList<String> manifestEntries, String path, AssetManager assetManager) throws IOException {
+        String[] fileList = assetManager.list(path);
+        if (fileList.length > 0) {
+            // This is a folder, recursively add folder entries to the manifest.
+            for (String pathInFolder : fileList) {
+                addFolderEntriesToManifest(manifestEntries, path + "/" + pathInFolder, assetManager);
+            }
+        } else {
+            // This is a file, compute a hash and create a manifest entry for it.
+            manifestEntries.add(path + ":" + computeHash(assetManager.open(path)));
+        }
+    }
+
+    private static String computeHash(InputStream dataStream) throws IOException {
+        MessageDigest messageDigest = null;
+        DigestInputStream digestInputStream = null;
+        try {
+            messageDigest = MessageDigest.getInstance("SHA-256");
+            digestInputStream = new DigestInputStream(dataStream, messageDigest);
+            byte[] byteBuffer = new byte[1024 * 8];
+            while (digestInputStream.read(byteBuffer) != -1);
+        } catch (NoSuchAlgorithmException e) {
+            // Should not happen.
+            throw new RuntimeException("Unable to compute hash of update contents.", e);
```

Comment:
This will probably kill the application if it happens. We could add `NoSuchAlgorithmException` to the list of thrown exceptions and handle it properly in the caller code. I think for now we only handle `IOException` there.